### PR TITLE
Support building the documenation on ReadTheDocs

### LINKED ISSUE #1610: Build plain ERT documentation
We should build a plain version of the ERT documentation and host it on Read The Docs or a similar service.

### PR DESCRIPTION
**Pull request by verveerpj**
**Issue**
Build and publish the documentation on ReadTheDocs  (issue #1610) 

**Approach**
Add a configuration file for ReadTheDocs.

Some issues to be done:
- ReadTheDocs must install ERT, otherwise the documentation will not build. This was straightforward, but failed promptly after we merged libres in. **Edit**: that was fixed easily.
- The documentation has a generation step for the communication protocol, ~~it remains to be seen if that can be  done on ReadTheDocs.~~ **Edit**: now done in sphinx extension that generates the doc pages.
- ReadTheDocs requires an account, we need to determine what that should be.

### CODE REVIEW

**Comment by verveerpj** [comment]
Getting the communication protocol section proves to be tricky. When generating the documentation currently we do the following:
1. Set a environment variable to point to a new directory and run some tests with python that generate some json files.
2. With the same environment variable set, run sphinx, which then uses a custom extension to generate documenation from the json files.

Step 2. can be easily done on ReadTheDocs by replacing the environment variable with some fallback location. Step 1 involves running pytest, which is not really doable on ReadTheDocs.

Possible solutions:
1. Generate the json files and check them in, possibly with some test to check if they are still the same when re-generated. Not particularly elegant, but should work.
2. Change the way we generate those, so that we can call that from the sphinx extension.
3. Ignore that part of the documentation.

**Comment by jondequinor** [comment]
> 2. Change the way we generate those, so that we can call that from the sphinx extension.

What options do we have here? We don't really need pytest to generate these. If you can call arbitrary ERT python, we should be good.

**Comment by verveerpj** [comment]
> > 1. Change the way we generate those, so that we can call that from the sphinx extension.
> 
> What options do we have here? We don't really need pytest to generate these. If you can call arbitrary ERT python, we should be good.

That would need to be done by sphinx, so the ErtNarratives sphinx extension would be the logical place.

**Comment by verveerpj** [comment]
Does generating the narratives json files have to involve running an ensemble? I guess so, which makes me somewhat uncomfortable since it is kind of uncontrolled where it will run, so all kind of failures could occur, such as websocket problems...

**Comment by jondequinor** [comment]
> Does generating the narratives json files have to involve running an ensemble?

No. You can produce the serialized narratives like this:

```python
import importlib
import tests.narratives


for narrative in tests.narratives.__all__:
    print(getattr(tests.narratives, narrative).json())
```

And then just dump that to some folder. One file per narrative.

**Comment by verveerpj** [comment]
> > Does generating the narratives json files have to involve running an ensemble?
> 
> No. You can produce the serialized narratives like this:
> 
> ```python
> import importlib
> import tests.narratives
> 
> 
> for narrative in tests.narratives.__all__:
>     print(getattr(tests.narratives, narrative).json())
> ```
> 
> And then just dump that to some folder. One file per narrative.

Yes, I found that out just before and used it to dump the files. For now I went for a solution where the files are committed and the tests are extended to check if they are up to date. If not, the assert message contains the command needed to update the files.

So, no generating during the sphinx run for  now, I find it a bit weird to run code form the tests folder during doc generation.

But that can be changed if there is a consensus how to do it best.

**Comment by verveerpj** [comment]
The current location of the narratives code in tests is cumbersome. Currently we are moving the tests, so the current approach fails immediately since the tests cannot find the documentation. This would not occur if we call the narrative code from a sphinx extension, since then we can use the tests located in the repository root, but the approach seems flaky. Should we not move the core narrative code into the main ERT source tree? If we consider it important for both documentation and testing, maybe it should live there.

**Comment by jondequinor** [comment]
> The current location of the narratives code in tests is cumbersome.

Agree, but that's why the json files existed. Not a satisfying solution in relation to this issue, since you need to run the tests. No direct dependence on the testing code, but a very painful indirect one.

> Should we not move the core narrative code into the main ERT source tree? If we consider it important for both documentation and testing, maybe it should live there.

If you're okay with that, then I am okay with that. I think that generally schemas are part of non-test code, and this is a schema-like thing.

In my mind the tests and the schemas had to be close together, but I can't really see why anymore.

> Edit: I implemented that approach and that cleans everything nicely.

Cool 👏

**Review by jondequinor — APPROVED** [review]
> The current location of the narratives code in tests is cumbersome.

Agree, but that's why the json files existed. Not a satisfying solution in relation to this issue, since you need to run the tests. No direct dependence on the testing code, but a very painful indirect one.

> Should we not move the core narrative code into the main ERT source tree? If we consider it important for both documentation and testing, maybe it should live there.

If you're okay with that, then I am okay with that. I think that generally schemas are part of non-test code, and this is a schema-like thing.

In my mind the tests and the schemas had to be close together, but I can't really see why anymore.

> Edit: I implemented that approach and that cleans everything nicely.

Cool 👏

File: tests/narrative_proxy.py [hunk 1]
```diff
@@ -6,7 +6,10 @@
...
 import websockets
 
-from tests.narrative import InteractionDirection, _Narrative
+from ert_shared.ensemble_evaluator.narratives.narrative import (
+    InteractionDirection,
+    _Narrative,
+)
 
 
 class NarrativeProxy(object):
```
Comment (jondequinor) [inline tests/narrative_proxy.py:15]: Should probably move this file as well. To `ert_shared.ensemble_evaluator.narratives.proxy`

**Pull request merged**